modules/distributor/forwarder.go

```diff
@@ -0,0 +1,311 @@
+				if tm.shouldUpdate(queueSize, workerCount) {
+					go func() {
+						// shutdown the queue manager
+						// this will block until all workers have finished and the queue is drained
+						if err := tm.shutdown(); err != nil {
+							level.Error(log.Logger).Log("msg", "error shutting down queue manager", "tenant", tenantID, "err", err)
+						}
+					}()
+					delete(f.queueManagers, tenantID)
+					f.queueManagers[tenantID] = newQueueManager(tenantID, queueSize, workerCount, f.forwardFunc)
+				}
+			}
+			f.mutex.Unlock()
+		case <-f.shutdown:
+			return
+		}
+	}
+}
+
+// TODO: Collecting in intervals instead of synchronously can hide spike loads
+//  Measure impact of updating the metric synchronously and consider changing the implementation
+func (f *forwarder) watchMetrics() {
+	collectInterval := 5 * time.Second
+	t := time.NewTicker(collectInterval)
+
+	for {
+		select {
+		case <-t.C:
+			f.mutex.Lock()
+			for tenant, qm := range f.queueManagers {
+				metricForwarderQueueLength.WithLabelValues(tenant).Set(float64(qm.queueLen()))
+			}
+			f.mutex.Unlock()
+		case <-f.shutdown:
+			return
+		}
+	}
+}
+
+func (f *forwarder) start(_ context.Context) error {
+	go f.watchOverrides()
+
+	go f.watchMetrics()
+
+	return nil
+
+}
+
+func (f *forwarder) stop(_ error) error {
+	close(f.shutdown)
+	var errs []error
+	for _, tm := range f.queueManagers {
+		if err := tm.shutdown(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return multierr.Combine(errs...)
+}
+
+// queueManager manages a single tenant's queue
+type queueManager struct {
+	// wg is used to wait for the workers to drain the queue while stopping
+	wg sync.WaitGroup
+
+	tenantID         string
+	workerCount      int
+	workerAliveCount *atomic.Int32
+	queueSize        int
+	reqChan          chan *request
+	fn               forwardFunc
+	workersCloseCh   chan struct{}
+
+	readOnly *atomic.Bool
+}
+
+func newQueueManager(tenantID string, queueSize, workerCount int, fn forwardFunc) *queueManager {
+	m := &queueManager{
+		tenantID:         tenantID,
+		workerCount:      workerCount,
+		queueSize:        queueSize,
+		workerAliveCount: atomic.NewInt32(0),
+		reqChan:          make(chan *request, queueSize),
+		fn:               fn,
+		workersCloseCh:   make(chan struct{}),
+		readOnly:         atomic.NewBool(false),
+	}
+
+	m.startWorkers()
+
+	return m
+}
+
+// pushToQueue a trace to the queue
+// if the queue is full, the trace is dropped
+func (m *queueManager) pushToQueue(ctx context.Context, req *request) error {
+	if m.readOnly.Load() {
+		return fmt.Errorf("queue is read-only")
+	}
+
+	select {
+	case m.reqChan <- req:
+		// TODO: Record some metric
+	case <-ctx.Done():
+		return fmt.Errorf("failed to pushToQueue traces to tenant %s queue: %w", m.tenantID, ctx.Err())
+	default:
+		// Fail fast if the queue is full
+		return fmt.Errorf("failed to pushToQueue traces to tenant %s queue: queue is full", m.tenantID)
```

Comment:
I'm probably misunderstanding channels and `select`, but doesn't this mean we never wait for `ctx.Done()`? What if `ctx` has a 15 second deadline, shouldn't we wait 15 seconds as the channel might get emptied by then?

Reply:
We don't wait, no. `<-ctx.Done()` will only be entered if the context is canceled by the time it arrives to the forwarder.

The idea here is to _fail fast_ if the queue is full, and not impact the ingest path is metrics-generators are being slow. If we wait for the timeout to be cancelled, latency would be propagated upstream. Which is something we want to hide (as I understand).

Reply:
Okay that makes sense, priority here is to not block :+1: